scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ClassDefinitionSetWithExtensionMethods.scala

```diff
@@ -0,0 +1,102 @@
+package pl.touk.nussknacker.engine.extension
+
+import cats.data.ValidatedNel
+import pl.touk.nussknacker.engine.ModelData
+import pl.touk.nussknacker.engine.api.Documentation
+import pl.touk.nussknacker.engine.api.generics.{GenericFunctionTypingError, MethodTypeInfo, Parameter}
+import pl.touk.nussknacker.engine.api.process.ClassExtractionSettings
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypingResult, Unknown}
+import pl.touk.nussknacker.engine.definition.clazz.{
+  ClassDefinition,
+  ClassDefinitionSet,
+  FunctionalMethodDefinition,
+  MethodDefinition
+}
+import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
+
+import java.lang.reflect.Method
+
+class ClassDefinitionSetWithExtensionMethods(set: ClassDefinitionSet, settings: ClassExtractionSettings) {
+  private val stringClass: Class[String] = classOf[String]
+  private val stringTyping: TypingResult = Typed.genericTypeClass(stringClass, Nil)
+  // We cannot have a class as the key because of `visibleMembersPredicate` e.g. Cast.castTo may be accessible for many classes.
+  // We only added statically defined methods because we needed a typingFunction which has access to some state.
+  // This functionality can be easily extended using dynamic methods discovery what you only need to do is:
+  // - Take all extension classes from class: ExtensionMethods, and then filter out those classes that are statically defined e.g. using annotation.
+  // - Filter out static and scala varargs methods (Method.isStatic and ClassDefinitionExtractor.javaVersionOfVarArgMethod)
+  // - Then invoke ClassDefinitionExtractor.extractMethod (you can easily build ClassDefinitionExtractor using settings).
+  // - After this we should filter out hidden members using ClassDefinitionExtractor.filterHiddenParameterAndReturnType.
+  // - And calculated definitions should be combined with static definitions.
+  private val extensionMethodsMap: Map[Method, Map[String, List[MethodDefinition]]] =
+    createStaticallyDefinedMethodsMap(set)
+
+  val value = new ClassDefinitionSet(
```

Comment:
Why we have set and value and both are public? :)

Reply:
Also, please move `value` and `extensionMethodsMap` initialization outside of class body, to factory method (apply/create). The general rule is to not have too much code inside the constructors. Body of class is the same as constructor in java.